app/Http/Controllers/Webapp/PostController.php

```diff
@@ -0,0 +1,1890 @@
+use Nayjest\Grids\Components\Base\RenderableRegistry;
+use Nayjest\Grids\Components\HtmlTag;
+use Nayjest\Grids\Components\Laravel5\Pager;
+
+use HTML;
+use Nayjest\Grids\SelectFilterConfig;
+use Psy\Exception\BreakException;
+
+class PostController extends NeedAuthController
+{
+    public function __construct(Route $route, Request $request) {
+        parent::__construct();
+
+        $this->activityLog = new Activity();
+        $this->activityLog->setSession($this->currentSession);
+
+        $actionName = explode("\\", $route->getActionName());
+        $actionName = $actionName[count($actionName)-1];
+
+        $this->activityLog->setAction($actionName);
+        if(($postId = $request->get('post_id', false)) !== false) {
+            $this->activityLog->setPostId($postId);
+        }
+    }
+
+    public function __destruct() {
+        $exceptActions = [
+            'PostController@getIndex',
+            'PostController@postStashChange',
+//            'PostController@postLock',
+            'PostController@postAutoSave'
+        ];
+
+        if(array_search($this->activityLog->action, $exceptActions) !== false) {
+            return;
+        }
+
+        $this->activityLog->save();
+    }
+
+    public function getIndex(Request $request) {
+        $query = DB::getDoctrineConnection()->createQueryBuilder();
+        $query->select([
+                'p.id',
+                'title',
+                'published_status',
+                'target_post_id',
+                'p.id_post_old',
+                'site_url',
+                'fullname',
+                'name',
+                'keyword',
+                'p.word_count',
+                'type',
+                'p.user_id',
+                'p.is_delete'
+            ])
+            ->from('posts', 'p')
+            ->leftJoin('p', 'sites', 's', 'p.site_id=s.id AND s.status = 1')
+            ->leftJoin('p', 'post_meta', 'pm', 'p.id = pm.post_id')
+            ->leftJoin('pm', 'keywords', 'k', 'pm.keyword_id = k.id')
+            ->innerJoin('p', 'users', 'u', 'p.user_id=u.id');
+
+        $userQueryWhere = '';
+        if(!$this->currentUser->is(User::ADMIN)) {
+            $userQueryWhere = ' AND is_delete = 0 ';
+        }
+        if($this->currentUser->is(User::EDITOR) || $this->currentUser->is(User::CONTRIBUTOR)) {
+            $userQueryWhere .= ' AND p.user_id = '.$this->currentUser->id;
+        }
+        if($request->get('keyword')) {
+            $_kw = urldecode($request->get('keyword'));
+            $queryWhere = '((p.site_id > 0 AND s.id > 0) OR (p.site_id = 0 OR p.site_id IS NULL)) AND keyword=\'' . mysqli_real_escape_string(new \mysqli(env('DB_HOST'), env('DB_USERNAME'), env('DB_PASSWORD'), env('DB_DATABASE'), env('DB_PORT')), $_kw) . '\' AND k.type = 1'.$userQueryWhere;
+        } else if(empty($request->all())) {
+            $queryWhere = '((p.site_id > 0 AND s.id > 0) OR (p.site_id = 0 OR p.site_id IS NULL)) ' . $userQueryWhere . ' GROUP BY p.id';
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Define a constant instead of duplicating this literal " GROUP BY p.id" 3 times. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://sonar.relipa.blog/coding_rules#rule_key=php%3AS1192)
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Define a constant instead of duplicating this literal "((p.site_id > 0 AND s.id > 0) OR (p.site_id = 0 OR p.site_id IS NULL)) " 4 times. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://sonar.relipa.blog/coding_rules#rule_key=php%3AS1192)